### DISCUSSION TITLE
Programmatically creating tcp appender

### DISCUSSION THREAD

**Discussion by GabrielJuliao** [logfellow/logstash-logback-encoder#918] [Q&A] [answered]
Is there a way to create a TCP appender programmatically?

**Comment by philsttr** [comment, marked as answer]
Yes, you would follow the same approach as [the answers to this Stack Overflow question regarding logback programmatic configuration](https://stackoverflow.com/questions/16910955/programmatically-configure-logback-appender), but instead of constructing and configuring those other appenders, you would construct and configure a [`LogstashTcpSocketAppender`](https://github.com/logfellow/logstash-logback-encoder/blob/main/src/main/java/net/logstash/logback/appender/LogstashTcpSocketAppender.java).  The xml element names seen in the [readme](https://github.com/logfellow/logstash-logback-encoder/blob/main/README.md) correlate to setter/adder methods on the appender and its nested objects.

**Comment by GabrielJuliao** [comment]
Thanks @philsttr phillsttr! How do I get LogstashTcpSocketAppender from the context?

I'm able to get logback core SSL appender like this: 

`LoggerContext ctx = (LoggerContext) LoggerFactory.getILoggerFactory();
Appender<ILoggingEvent> appender = ctx.getLogger("ROOT").getAppender("SSL_APPENDER");
`
but the return type of getAppender is `Appender<ILoggingEvent>`

basically what I need to do is stop and start the appender to reload the Keystore/Truststore, I can do this with SSLSocketAppender.

**Reply by philsttr** [reply]
Assuming you have added a `LogstashTcpSocketAppender` programmatically or via logback.xml, you would call `getAppender` on the logger to which you added the appender, passing the name of the appender, and then downcast the returned value to `LogstashTcpSocketAppender`.

If you have configured a `LogstashTcpSocketAppender` in logback xml, then the name is specified in the `name` attribute of the `appender` xml element.
If you have configured a `LogstashTcpSocketAppender` programmatically, then the name is specified by calling `appender.setName`

**Reply by GabrielJuliao** [reply]
This is my code to reload the appender:

```java
    @GetMapping("/restart")
    public ResponseEntity<?> restart() {
        LoggerContext ctx = (LoggerContext) LoggerFactory.getILoggerFactory();
        Appender<ILoggingEvent> appender = ctx.getLogger("ROOT").getAppender("SSL_APPENDER");
        if (Objects.nonNull(appender)) {
            logger.info("Stopping appender");
            appender.stop();
            try {
                logger.info("Starting appender");
                appender.start();
            } catch (Exception e) {
                logger.error("Error starting appender: " + e.getMessage(), e);
                return ResponseEntity.status(500).build();
            }
            return ResponseEntity.ok().build();
        }

        logger.info("Appender not found");
        return ResponseEntity.notFound().build();
    }
```

I'm trying to downcast, but getting inconvertible types...

**Reply by philsttr** [reply]
what does the definition of SSL_APPENDER look like?  Is it defined as a `LogstashTcpSocketAppender`?

**Reply by GabrielJuliao** [reply]
`
<?xml version="1.0" encoding="UTF-8"?>
<configuration debug="false">

    <property name="LOG_PATTERN" value="%d{yyyy-MM-dd HH:mm:ss.SSSZZ} %5p ${PID:- } --- [%15.15t] %-40.40logger{39} : %m%n"/>
    <appender name="CONSOLE_APPENDER" class="ch.qos.logback.core.ConsoleAppender">
        <encoder>
            <Pattern>${LOG_PATTERN}</Pattern>
        </encoder>
    </appender>

    <appender name="SSL_APPENDER" class="net.logstash.logback.appender.LogstashAccessTcpSocketAppender">
        <destination>localhost:8000</destination>
        <reconnectionDelay>2 seconds</reconnectionDelay>
        
        <ssl>
            <keyStore>
                <location>${C_TCP_LOGGING_KEYSTORE}</location>
                <password>${C_TCP_LOGGING_KEYSTORE_PASSWORD}</password>
            </keyStore>
            <trustStore>
                <location>${C_TCP_LOGGING_TRUSTSTORE}</location>
                <password>${C_TCP_LOGGING_TRUSTSTORE_PASSWORD}</password>
            </trustStore>
        </ssl>
        <encoder>
            <Pattern>${LOG_PATTERN}</Pattern>
        </encoder>
    </appender>

    <root level="INFO">
        <appender-ref ref="SSL_APPENDER"/>
        <appender-ref ref="CONSOLE_APPENDER"/>
    </root>

</configuration>
`

**Reply by GabrielJuliao** [reply]
I was able to do the downcast but for some reason it does not refresh keystore/truststore even if you create new ones. The code above works for SSLSocketAppender, and keystore/truststore are refreshed.

**Reply by GabrielJuliao** [reply]
@philsttr Thanks for you help! I was able to reload the Keystore by simply stopping the appender then setting the socket factory to null and starting again!!!
